### PR TITLE
[4.0] Removing artificial sharding from com_finder

### PR DESCRIPTION
**Pull request by Hackwar**
Finder has an artificial sharding, which splits up the links between terms and content items onto 16 seperate tables. This PR removes that sharding. It makes the code extremely complex and bloats it unnecessarily and it also slows down the whole finder system. Indexing the original sample data of Joomla 4.0 with the current finder indexer takes about 2:40 minutes on my system. With these changes, we are already down to 1:30 minute. 

This PR also fixes a few other bugs that prevented com_finder from working in Joomla 4.0 right now and I standardised all the language fields to be 7 characters long. While I plan to provide more PRs with further improvements to make com_finder more usable, this PR is in itself complete and can be merged like this.

## How to test
Install 4.0-dev and go to the backend of Smart Search. Enable the Smart Search Plugin and start the indexing process. Notice that it takes a long time (preferably measure the time with a stopwatch). Go to the frontend and open the Smart Search menu item. See that it throws errors and doesn't work right now.
Apply this PR and clear the index in the backend. Start the re-indexing and notice that it finishes quicker. Go to the frontend and see that that menu item no longer throws errors and that all features work like expected.

When typing in a phrase and then pressing enter, sending the request to the server, a message "error" is displayed before the page is reloaded both before and after applying the patch. I'm still looking for the source of that issue.

I want to thank [cloudaccess.net](https://www.cloudaccess.net/) for supporting these changes. Without their generous help, this wouldn't be possible.

### CODE REVIEW

**Comment by brianteeman** [comment]
As there is  administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql
I assume that there should also be an equivalent postgres file?

**Comment by Hackwar** [comment]
I would need help with the postgres stuff. I don't have postgres on my system and would need someone to provide such a file and test that part.

**Comment by Hackwar** [comment]
@alikon Could you help me with adding the right migration script for postgres?

**Comment by alikon** [comment]
yes i'll do it

**Comment by Hackwar** [comment]
Thank you!

**Comment by infograf768** [comment]
Will test. Please let me know, apart from the indexing speed (and obvious code corrections), what I can expect, as a multilingual user, from this PR?

**Comment by Hackwar** [comment]
At this moment, there is no change in terms of multilingual users. In order to keep these changes testable, I'm restricting my changes to one thing at a time. So this only removes the sharding and gets it all working again in 4.0. The language changes are coming in the next few steps.

**Comment by alikon** [comment]
made a pr https://github.com/Hackwar/joomla-cms/pull/22
unfortunately without this one https://github.com/joomla/joomla-cms/pull/19707 merged 
unable to test/work

**Comment by Hackwar** [comment]
Merged

**Comment by brianteeman** [comment]
I was unable to replicate the error that you mentioned before applying this PR

After applying the PR the indexer did not work at all

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 7716cd6ca287d28c8586a9fa16466316e28aa4a3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20185">issues.joomla.org/tracker/joomla-cms/20185</a>.</sub>

**Comment by Hackwar** [comment]
The current 4.0-dev branch fails in displaying the frontend when opening the smart search component and having debug mode enabled with the following error message:
1055 'j4search2.t.parent_id' isn't in GROUP BY 
This has been fixed in this PR.

I have updated the branch with the latest changes from 4.0-dev and checked it again. It does work. Could it be that you did not apply the database changes, that this introduces? I would be happy if you could test this again.

**Comment by Hackwar** [comment]
@brianteeman Did you see my last message? Could you check your test again, especially with the database changes?

**Comment by brianteeman** [comment]
Sorry with JAB and JDFR it slipped my radar - thanks for the reminder I will put it on the todo for tomorrow

**Comment by Hackwar** [comment]
Just to list all current PRs for Finder/Smart Search:
* #20185 [4.0] Removing artificial sharding from com_finder
* #20384 [4.0] Removing terms tuples from com_finder
* #20391 [4.0] Using language specific tokeniser and stemmer for com_finder

**Comment by brianteeman** [comment]
@Hackwar no change from my previous report

**Comment by Hackwar** [comment]
Can you see what the ajax call from the indexer returns as error?

**Comment by brianteeman** [comment]
```
core.min.js?bcb349ab2b37648abc4c1018be364433:1 GET http://127.0.0.1/cms4/administrator/index.php?option=com_finder&tmpl=component&format=json&task=indexer.batch&ecfbe9f00cc7263cb121fc56d4885d68=1 500 (Internal Server Error)
e.request @ core.min.js?bcb349ab2b37648abc4c1018be364433:1
i @ indexer.min.js?bcb349ab2b37648abc4c1018be364433:1
l @ indexer.min.js?bcb349ab2b37648abc4c1018be364433:1
onSuccess @ indexer.min.js?bcb349ab2b37648abc4c1018be364433:1
n.onreadystatechange @ core.min.js?bcb349ab2b37648abc4c1018be364433:1
XMLHttpRequest.send (async)
e.request @ core.min.js?bcb349ab2b37648abc4c1018be364433:1
i @ indexer.min.js?bcb349ab2b37648abc4c1018be364433:1
e.finderIndexer @ indexer.min.js?bcb349ab2b37648abc4c1018be364433:1
(anonymous) @ indexer.min.js?bcb349ab2b37648abc4c1018be364433:1
indexer.min.js?bcb349ab2b37648abc4c1018be364433:1 Uncaught ReferenceError: json is not defined
    at m (indexer.min.js?bcb349ab2b37648abc4c1018be364433:1)
    at onError (indexer.min.js?bcb349ab2b37648abc4c1018be364433:1)
    at XMLHttpRequest.n.onreadystatechange (core.min.js?bcb349ab2b37648abc4c1018be364433:1)
```

**Comment by Hackwar** [comment]
Sorry to bother you, but can you also see what the specific response is from the ajax call? So far we only know that it throws a 500 error... In best case, enable debugging first.

**Comment by brianteeman** [comment]
```
core.js?35800ffe54a73a00bb236e36e4b1839f:890 GET http://127.0.0.1/cms4/administrator/index.php?option=com_finder&tmpl=component&format=json&task=indexer.batch&ecfbe9f00cc7263cb121fc56d4885d68=1 500 (Internal Server Error)
Joomla.request @ core.js?35800ffe54a73a00bb236e36e4b1839f:890
getRequest @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:29
handleResponse @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:68
onSuccess @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:36
xhr.onreadystatechange @ core.js?35800ffe54a73a00bb236e36e4b1839f:876
XMLHttpRequest.send (async)
Joomla.request @ core.js?35800ffe54a73a00bb236e36e4b1839f:890
getRequest @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:29
initialize @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:25
Joomla.finderIndexer @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:164
(anonymous) @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:169
indexer.js?35800ffe54a73a00bb236e36e4b1839f:107 Uncaught ReferenceError: json is not defined
    at handleFailure (indexer.js?35800ffe54a73a00bb236e36e4b1839f:107)
    at onError (indexer.js?35800ffe54a73a00bb236e36e4b1839f:39)
    at XMLHttpRequest.xhr.onreadystatechange (core.js?35800ffe54a73a00bb236e36e4b1839f:879)
handleFailure @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:107
onError @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:39
xhr.onreadystatechange @ core.js?35800ffe54a73a00bb236e36e4b1839f:879
XMLHttpRequest.send (async)
Joomla.request @ core.js?35800ffe54a73a00bb236e36e4b1839f:890
getRequest @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:29
handleResponse @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:68
onSuccess @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:36
xhr.onreadystatechange @ core.js?35800ffe54a73a00bb236e36e4b1839f:876
XMLHttpRequest.send (async)
Joomla.request @ core.js?35800ffe54a73a00bb236e36e4b1839f:890
getRequest @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:29
initialize @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:25
Joomla.finderIndexer @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:164
(anonymous) @ indexer.js?35800ffe54a73a00bb236e36e4b1839f:169

```

**Comment by Hackwar** [comment]
That is still the JS trace, not the HTML error that Joomla is returning.

**Comment by infograf768** [comment]
Tested here on a clean patched 4.0-dev on Firefox
I could index default sample data without any error.

Search via the module in frontend does work fine although I have a small js error
`TypeError: t is null[Learn More]
finder.min.js:1:985`

This error is only present when searching from the module, i.e. once the the results are displayed and one searches from the search field displayed on top of the results, the error is not thrown.

**Comment by infograf768** [comment]
After enabling debug, I get a different js error
`TypeError: advanced is null[Learn More]
finder.js:62:1`

**Comment by Hackwar** [comment]
I did not change anything in the output or in the JS. Those errors have been present before and are stuff for yet another PR.

**Comment by carlitorweb** [comment]
I have tested this item :red_circle: unsuccessfully on aec1df665dcd701a6dc9f97a3349aceb30234ed9<br /><br /> 
> Start the indexing process. Notice that it takes a long time. Go to the frontend and open the Smart Search menu item. See that it throws errors and doesn't work right now.

I confirm this

> Apply this PR and clear the index in the backend. Start the re-indexing and notice that it finishes quicker. Go to the frontend and see that that menu item no longer throws errors and that all features work like expected.

After applying the PR the indexer did not work at all<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20185">issues.joomla.org/tracker/joomla-cms/20185</a>.</sub>

**Comment by Hackwar** [comment]
Please make sure that the db changes are applied.Am 24.05.2018 22:22 schrieb Carlos Rodriguez <notifications@github.com>:I have tested this item 🔴 unsuccessfully on aec1df6>  start the indexing process. Notice that it takes a long time. Go to the frontend and open the Smart Search menu item. See that it throws errors and doesn't work right now.
I confirm this

Apply this PR and clear the index in the backend. Start the re-indexing and notice that it finishes quicker. Go to the frontend and see that that menu item no longer throws errors and that all features work like expected.

After applying the PR the indexer did not work at allThis comment was created with the J!Tracker Application at issues.joomla.org/tracker/joomla-cms/20185.

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on aec1df665dcd701a6dc9f97a3349aceb30234ed9<br /><br />
> Please make sure that the db changes are applied

Sorry @Hackwar I not saw this


> Apply this PR and clear the index in the backend. Start the re-indexing and notice that it finishes quicker. Go to the frontend and see that that menu item no longer throws errors and that all features work like expected.


Now work, as described <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20185">issues.joomla.org/tracker/joomla-cms/20185</a>.</sub>

**Comment by carlitorweb** [comment]
Will be good if we can have this merged, for can test others com_finder PR you did.

**Comment by Hackwar** [comment]
@brianteeman can you again make sure that the DB changes have been applied and test this again?

@infograf768 is your test a successfull test or do you consider that as an error? 

If you could mark/update your tests, we could merge this. :wink:

**Comment by infograf768** [comment]
Before setting this to OK, I need the advice of @chrisdavenport as this does not only solves a bug but also changes the database structure of com_finder.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on aec1df665dcd701a6dc9f97a3349aceb30234ed9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20185">issues.joomla.org/tracker/joomla-cms/20185</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by carlitorweb** [comment]
@wilsonge @laoneo  can we have this merged? I will like test others com_finder PR, Which depend on this

File: components/com_finder/router.php [hunk 1]
```diff
@@ -9,118 +9,34 @@
 
 defined('_JEXEC') or die;
 
-use Joomla\CMS\Component\Router\RouterBase;
+use Joomla\CMS\Component\Router\RouterView;
```
Comment (chrisdavenport) [inline components/com_finder/router.php]: Removing sharding should not have any effect on routing, so why are changes being made to this class here?  I think these changes should they be in a separate PR.

**Comment by chrisdavenport** [comment]
@infograf768  I haven't tested this PR, but I agree in principle with the intention.  I ran some tests a while back that convinced me that there should be a performance improvement in most cases.

Watch out for older versions of the database having columns with default values that are no longer valid, as this could cause the ALTER TABLE statements to fail.

File: components/com_finder/router.php (hunk 1 above)
Comment (Hackwar) [reply components/com_finder/router.php]: I actually don't really know why I refactored the router in this, too. If you insist on making that a separate PR, I can do that. On the other hand, this is a rather trivial change from my perspective...

**Comment by Hackwar** [comment]
I don't think that the old default values are an issue. I've proposed quite a lot of changes, which in the end requires rebuilding the index anyway. When we are done with all changes, I would consolidate all SQL changes into one file and prepend that with a few TRUNCATEs to clear the index.

**Comment by laoneo** [comment]
I couldn't really figure out what would be the consequences of this pr. What will the then not work as before when this pr get merged?

**Comment by Hackwar** [comment]
Everything works as before. We only use 15 db tables less. It speeds up indexing and makes the code a lot easier to read.

**Comment by Hackwar** [comment]
Should I now revert the router changes again or can this be merged the way it is?

**Comment by laoneo** [comment]
I just can't believe that removing 16 tables has not a single bit of consequences. Something must not work as before. As long as we do not know that, I don't suggest that we merge this pr.

**Comment by brianteeman** [comment]
@laoneo all that is happening is that an artificial separation of the data into 16 tables no longer happens and the data is just in a single table

**Comment by laoneo** [comment]
Why was that done in that way and now we can move it to one table. Don't get me wrong I'm not against it, just want to make sure we do not miss anything.

**Comment by Hackwar** [comment]
According to a discussion that I had with Louis Landry half a millenia ago, that splitting was meant to improve performance. According to him, inserting all those tens of thousands of rows into a single table would slow down the indexing process a lot, since the DB at some point has to catch up with all the inserts. Splitting this up into these 16 tables (16 because it allowed for easy calculations with hex values) was supposed to prevent that, because now the insertions would count for each table separately, thus increasing the throughput by the factor of 16. 

I don't want to judge if that actually was true at the time that finder was developed, but according to my tests and a few people that I talked to in the SQL IRC channel, that is not the case. Having these split tables actually is a big performance hit when indexing and requires up to 16 queries to get all search results from our system. This is a useless feature that no one needs, that makes our system slower, the code unmaintainable and which has to go as soon as possible. 

If you have a website that is so large that you really have issues here, then you can have a look at the built in sharding feature of MySQL. But I would honestly instead expect you to use a standalone search system like Elastic or Solr or similar.

**Comment by laoneo** [comment]
Thanks for explanation. That was the information I was looking for.

**Comment by chrisdavenport** [comment]
For the record, I had similar conversations with Louis and Rob many years ago and I seem to recall them saying that the decision to shard was based on real-world experience at the time.  Consequently, when Hannes originally suggested that sharding be removed I ran some tests of my own.  I modified the code to allow a variable number of shards (actually only a power of 2).  In all cases, averaging over many test runs, there was a performance improvement on indexing.  In theory, I would expect a performance improvement on retrieval too, but I didn't test that.

The performance improvements (relative to 16 shards) were are follows:
8 shards: 26.1%
4 shards: 35.4%
2 shards: 43.7%
1 shard (ie. no sharding): 44.4%
Tests were carried out on a real-world data set consisting of 1445 articles containing 454046 terms resulting in 1215490 links-terms map entries.

A maintainer should decide whether the router changes, which have nothing to do with sharding, should be in a separate PR as they need a different kind of testing.

**Comment by laoneo** [comment]
Yes please one pr per task.

**Comment by laoneo** [comment]
Thanks Chris for the explanation.

**Comment by Hackwar** [comment]
I've reverted the changes of the router and cleaned up the commits.

File: components/com_finder/helpers/html/query.php [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 defined('_JEXEC') or die;
 
+JLoader::register('FinderHelperLanguage', JPATH_ADMINISTRATOR . '/components/com_finder/helpers/language.php');
```
Comment (laoneo) [inline components/com_finder/helpers/html/query.php:12]: Is this still needed?

**Comment by laoneo** [comment]
Would be good if we can get another round of testing and then I think it is good to merge.

File: components/com_finder/helpers/html/query.php (hunk 2 above)
Comment (Hackwar) [reply components/com_finder/helpers/html/query.php:12]: Yes, otherwise you get a fatal error when using the advanced filters. And please don't ask me to put all these single line fixes to make com_finder run in 4.0 into separate PRs....

**Comment by Hackwar** [comment]
We have 2 successfull tests. Why do we need another round of tests now?

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on c80b97523a0ba6db1909357b671f709633b02cf1<br /><br />Tested again. Only the route change was removed from this PR, so I think will be same. Anyway, I tested again, and work as before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20185">issues.joomla.org/tracker/joomla-cms/20185</a>.</sub>

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+DROP TABLE #__finder_links_terms7;
+DROP TABLE #__finder_links_terms8;
+DROP TABLE #__finder_links_terms9;
+DROP TABLE #__finder_links_termsa;
+DROP TABLE #__finder_links_termsb;
+DROP TABLE #__finder_links_termsc;
+DROP TABLE #__finder_links_termsd;
+DROP TABLE #__finder_links_termse;
+DROP TABLE #__finder_links_termsf;
+
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql:28]: Changed `CHANGE COLUMN` to `CHANGE`, dropped `AFTER ...` and be on same line to make it work.

```
ALTER TABLE `#__finder_terms` CHANGE `language` `language` CHAR(7) NOT NULL DEFAULT '';
ALTER TABLE `#__finder_terms_common` CHANGE `language` `language` CHAR(7) NOT NULL DEFAULT '';
ALTER TABLE `#__finder_tokens` CHANGE `language` `language` CHAR(7) NOT NULL DEFAULT '';
ALTER TABLE `#__finder_tokens_aggregate` DROP COLUMN `map_suffix`;
ALTER TABLE `#__finder_tokens_aggregate` CHANGE `language` `language` CHAR(7) NOT NULL DEFAULT '';
ALTER TABLE `#__finder_links` CHANGE `language` `language` CHAR(7) NOT NULL DEFAULT '';
```

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	`term_id` INT(10) UNSIGNED NOT NULL,
+	`weight` FLOAT UNSIGNED NOT NULL,
+	PRIMARY KEY (`link_id`, `term_id`),
+	INDEX `idx_term_weight` (`term_id`, `weight`),
+	INDEX `idx_link_term_weight` (`link_id`, `term_id`, `weight`)
+)
+COLLATE='utf8mb4_general_ci'
+ENGINE=InnoDB
+;
+DROP TABLE #__finder_links_terms0;
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql:12]: Tables are not dropped and I don't know why.

Comment (carlitorweb) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql:12]: I did this operation manually, and work ok

Comment (Quy) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql:12]: Yes, but not via `Extensions > Manage > Database > Fix`

Comment (brianteeman) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-04-14.sql:12]: Iirc you can't do a drop that way

**Comment by laoneo** [comment]
@Hackwar can you please add the suggested changes to make the DB fixing work and then we are good to go here.

**Comment by brianteeman** [comment]
there is no reason to make database>fix work
That is a misuse of the functionality

**Comment by laoneo** [comment]
Ok, misinterpreted that comment. What about the suggestion from @Quy?

**Comment by Hackwar** [comment]
I'm hesitant to change the SQL again, mainly because I don't see how what is supposed to work how and more importantly because there are quite a few additional changes on the way for the finder tables. At one point we will have to make one PR that combines all these changes into one SQL file. Can we tackle that issue then? (There are 12 open PRs for finder from me, 4 of which have DB changes)

**Comment by laoneo** [comment]
Does the upgrade script work without the recommendations from @Quy?

**Comment by Quy** [comment]
With fresh install, it works. 
With upgrade/patch tester, it stalls at `Starting Indexer`.

**Comment by mbabker** [comment]
This has database schema changes.  Patch tester can't deal with database schema changes (you can apply the patch and run `Database > Fix` but you can't reverse the schema changes if you do that).  The component, as useful as it can be, is not a definitive way of identifying issues with patches.

**Comment by Hackwar** [comment]
Just to clarify @chrisdavenport s comment here: With the sharding removed, he had a performance **improvement** of 44%, **not** a performance reduction.

**Comment by Hackwar** [comment]
Thank you!

**Comment by wilsonge** [comment]
Thankyou for being so patient with this! Sorry it took so long to deal with!
